compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs

```diff
@@ -2047,39 +2047,34 @@ impl<'tcx> TypeErrCtxtExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
         let hir::ExprKind::Path(path) = arg.kind else {
             return;
         };
-        let expected_inputs = self.tcx.erase_late_bound_regions(*expected).inputs();
-        let found_inputs = self.tcx.erase_late_bound_regions(*found).inputs();
-        let both_tys = expected_inputs.iter().cloned().zip(found_inputs.iter().cloned());
+        let expected_inputs = self.tcx.instantiate_bound_regions_with_erased(*expected).inputs();
+        let found_inputs = self.tcx.instantiate_bound_regions_with_erased(*found).inputs();
+        let both_tys = expected_inputs.iter().copied().zip(found_inputs.iter().copied());
 
         let arg_expr = |infcx: &InferCtxt<'tcx>, name, expected: Ty<'tcx>, found: Ty<'tcx>| {
             let (expected_ty, expected_refs) = get_deref_type_and_refs(expected);
             let (found_ty, found_refs) = get_deref_type_and_refs(found);
 
             if infcx.can_eq(param_env, found_ty, expected_ty) {
                 if found_refs.len() == expected_refs.len()
-                    && found_refs.iter().zip(expected_refs.iter()).all(|(e, f)| e == f)
+                    && found_refs.iter().eq(expected_refs.iter())
                 {
                     name
                 } else if found_refs.len() > expected_refs.len() {
-                    if found_refs[..found_refs.len() - expected_refs.len()]
-                        .iter()
-                        .zip(expected_refs.iter())
-                        .any(|(e, f)| e != f)
-                    {
-                        // The refs have different mutability.
+                    let refs = &found_refs[..found_refs.len() - expected_refs.len()];
+                    if found_refs[..expected_refs.len()].iter().eq(expected_refs.iter()) {
```

Comment:
By changing it from `zip` to `eq`, now I can't rely on stopping at the shortest one, so I need this second sub-slicing of `found_refs`. Deduplicated the sub-slicing withing the two if branches though.